### PR TITLE
Add toMatchSnapshot as an override matcher

### LINKED ISSUE #45: [Matcher]: `toMatchSnapshot` for enzyme comp
**`jest-enzyme` ONLY**

It would be real cool if we could support enzyme with snapshot testing in jest. Since we rejected it from living in the enzyme repo (https://github.com/airbnb/enzyme/pull/541), this seems like the natural place.

Of course we would have to fall back to the original `toMatchSnapshot` if a non-enzyme wrapper was given.

Basic thoughts:

``` js
var originalToMatchSnapshot;
toMatchSnapshot(enzymeWrapper) {
  if (notEnzyme(enzymeWrapper)) {
    return originalToMatchSnapshot(enzymeWrapper);
  }

  var snapshotable = convertToCompatibleJSON(enzymeWrapper);
  return originalToMatchSnapshot(snapshotable);
}
```

### PR DESCRIPTION
**Pull request by blainekasten**
Resolve #45 

The concept here is to override jests `toMatchSnapshot` matcher when an enzyme wrapper is given to it. The hope here is that Enzyme is easily compatible and stable with use of `toMatchSnapshot`. Currently people either have to pass the enzyme object to the matcher (which is terrible) or they have to use react-test-utils, which isn't ideal when someone is bought into Enzyme.

The current implementation has me a bit worried though. We're currently depending on a global state from jest that isn't publically supported. I'm also a bit worried that enzyme-to-json isn't the right solution.

I would love some feedback from those who are involved with these projects what they think of this.

cc @cpojer @dmitriiabramov @aweary @lelandrichardson

### CODE REVIEW

File: packages/jest-enzyme/src/toMatchSnapshot.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import wrapperToJSON from 'enzyme-to-json';
+
+const GLOBAL_STATE = Symbol.for('$$jest-matchers-object');
+const originalToMatchSnapshot = global[GLOBAL_STATE].matchers.toMatchSnapshot;
```
Comment (blainekasten) [inline packages/jest-enzyme/src/toMatchSnapshot.js]: specifically, is this safe? Is there a better way to get access to this reference?

File: packages/jest-enzyme/src/toMatchSnapshot.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+function isEnzyme(wrapper:any): boolean{
+  if (typeof wrapper !== 'object' || Array.isArray(wrapper)) {
+    return false;
+  }
+
+  return !!(wrapper.node && Array.isArray(wrapper.nodes) && wrapper.complexSelector);
```
Comment (blainekasten) [inline packages/jest-enzyme/src/toMatchSnapshot.js]: is this a safe enough check?

File: packages/jest-enzyme/src/toMatchSnapshot.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+export default function toMatchSnapshot(maybeEnzymeWrapper:any) {
+  let snapshotableValue = maybeEnzymeWrapper;
+  if (isEnzyme(maybeEnzymeWrapper)) {
+    snapshotableValue = wrapperToJSON(maybeEnzymeWrapper);
```
Comment (blainekasten) [inline packages/jest-enzyme/src/toMatchSnapshot.js]: is `enzyme-to-json` the right solution?

**Comment by dmitriiabramov** [comment]
i'm a little worried about overriding the matcher. would it be better to have something like `.toMatchEnzymeSnapshot()`?

**Comment by blainekasten** [comment]
@dmitriiabramov I was hoping the override would be cool as it would be easier for people to transition enzyme <--> react-test-utils

File: packages/jest-enzyme/src/toMatchSnapshot.js (hunk 1 above)
Comment (dmitriiabramov) [inline packages/jest-enzyme/src/toMatchSnapshot.js]: yeah. you shouldn't do this. Use get/set instead https://github.com/facebook/jest/blob/master/packages/jest-matchers/src/index.js#L183-L187

**Comment by dmitriiabramov** [comment]
yeah, but at the same time it increases the complexity, cause now different kind of objects will produce different kind of serialization

**Comment by cpojer** [comment]
yeah please don't overwrite `toMatchSnapshot`. @dmitriiabramov shall we protect the original matchers from Jest and guard against overwriting them?

**Comment by lelandrichardson** [comment]
Can we just add a better `toJSON` method on the enzyme wrapper which will work with jest's pretty printing. @cpojer @blainekasten

**Comment by blainekasten** [comment]
We had a PR for that once and @ljharb turned it down.

**Comment by ljharb** [comment]
(relevant PR: https://github.com/airbnb/enzyme/pull/541#issuecomment-239857292)

**Comment by lelandrichardson** [comment]
This is the OR that @blainekasten is referencing: https://github.com/airbnb/enzyme/pull/541

Couldn't all this be solved by just having:

```js
toJSON() {
  return this.node;
}
```

I'd be in favor of a PR that used `toJSON()` (rather than `json()`) and semantically represented the serialization of the wrapped node rather than the wrapper itself. I think in this case jest's snapshotting will produce something useful since it's a real react element?

I'll play around with this if I get time later and put up a PR for discussion.

**Comment by lelandrichardson** [comment]
And to be clear about my stance on snapshot testing:

I think snapshot testing can be misused, but does represent a good tradeoff for frontend testing and things where you want to detect regressions rather than assert on actual behavior. 

I was more skeptical about snapshotting early on but I have since changed my mind. I think that there are certain scenarios jest snapshotting is very useful for and i look at it as a valuable tool among others in a JS developer's "testing toolbox".

I also think that it's important for enzyme to work with jest's snapshot testing as I think that enzyme can actually make jest snapshot testing far more useful by making it easy to snapshot slices of a larger render tree, increasing signal to noise ratio.  There are also people trying to decouple jest snapshotting from jest and I think that's great.

**Comment by cpojer** [comment]
@lelandrichardson you got the right mindset. After all, snapshot testing is just one assertion out of 30 or so. It just happens to be a complex one with good UX. I believe it isn't actually that great yet; we can push the feature a lot further. A lot of it has also been blocked by improvements to the test renderer: once you can actually use the test-renderer to interact with your components and if you can snapshot partial components, then there will be a lot more value in snapshot testing for React. I think using enzyme for snapshot testing works pretty well, but of course it is still limited to the DOM, so it doesn't work for RN. @frantic is working on improvements for the test renderer currently.

We use snapshot testing pretty heavily for testing CLI output in Jest and it is a really good use-case and a great experience.

**Comment by blainekasten** [comment]
> yeah please don't overwrite toMatchSnapshot. @dmitriiabramov shall we protect the original matchers from Jest and guard against overwriting them?

That is probably a smart idea regardless. I accidentally overwrote `toContain` in the past and broke people's tests without realizing. 

> Couldn't all this be solved by just having:
> ```js
> toJSON() {
>   return this.node;
> }
> ```

Could it be that easy? If so, we should just do that. I'm thinking based on the implementation of [enzyme-to-json](https://github.com/adriantoine/enzyme-to-json/blob/master/src/shallow.js) that it appears a bit harder to implement.

> i'm a little worried about overriding the matcher. would it be better to have something like `.toMatchEnzymeSnapshot()`?

Since the consensus is to not override, this seems reasonable.

**Comment by ljharb** [comment]
fwiw I'm 100% in favor of adding a `toJSON` function, assuming `JSON.stringify()` produces meaningful output that way.

**Comment by blainekasten** [comment]
Closing this as it's not the right approach. I'm exploring another idea.

**Pull request closed without merging**